examples/nlp/text_generation_transformer_decoder.py

```diff
@@ -0,0 +1,229 @@
+    transformer_block = TransformerDecoder(
+        num_heads=NUM_HEADS,
+        intermediate_dim=FEED_FORWARD_DIM,
+    )
+    x = transformer_block(x) # no cross-attention
+  # output
+  outputs = keras.layers.Dense(VOCAB_SIZE)(x)
+  model = keras.Model(inputs=inputs, outputs=outputs)
+  loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+  model.compile(
+      "adam", loss=loss_fn,
+  )
+  return model
+
+"""
+## Training
+
+We use a callback to test out the text generation at every epoch. This callback uses
+top-k decoding.
+We also take a subset of the training data for quicker training since this is used as an
+example.
+"""
+
+class TopKTextGenerator(keras.callbacks.Callback):
+  """A callback to generate text from a trained model using top-k.
+  1. Feed some starting prompt to the model
+  2. Predict probabilities for the next token
+  3. Sample the next token and add it to the next input
+
+  Arguments:
+    max_tokens: Integer, the number of tokens to be generated after prompt.
+    start_tokens: List of integers, the token indices for the starting prompt.
+    tokenizer: WordPieceTokenizer object for de-tokenizing output.
+    top_k: Integer, sample from the top_k token predictions.
+    print_every: Integer, print after this many epochs.
+  """
+
+  def __init__(
+      self, max_tokens, start_tokens, tokenizer, top_k=10, print_every=1
+  ):
+    self.max_tokens = max_tokens
+    self.start_tokens = start_tokens
+    self.tokenizer = tokenizer
+    self.print_every = print_every
+    self.k = top_k
+
+  # top k sampling
+  def sample_from(self, logits):
+    logits, indices = tf.math.top_k(logits, k=self.k)
+    indices = np.asarray(indices).astype("int32")
+    preds = keras.activations.softmax(tf.expand_dims(logits, 0))[0]
+    preds = np.asarray(preds).astype("float32")
+    return np.random.choice(indices, p=preds)
+
+  def on_epoch_end(self, epoch, logs=None):
+    start_tokens = [_ for _ in self.start_tokens]
+    if (epoch + 1) % self.print_every != 0:
+      return
+    num_tokens_generated = 0
+    while num_tokens_generated <= self.max_tokens:
+      # pad sequences
+      pad_len = SEQ_LEN - len(start_tokens)
+      sample_index = len(start_tokens) - 1
+      if pad_len < 0:
+        x = start_tokens[:SEQ_LEN]
+        sample_index = SEQ_LEN - 1
+      elif pad_len > 0:
+        x = start_tokens + [0] * pad_len
+      else:
+        x = start_tokens
+      x = np.array([x])
+
+      y = self.model.predict(x)
+      sample_token = self.sample_from(y[0][sample_index])
+      start_tokens.append(sample_token)
+      num_tokens_generated += 1
+    txt = self.tokenizer.detokenize(start_tokens)
+    print(f"generated text: \n{txt}\n")
+
+start_prompt = "the"
+# Unpadded token sequence
+start_tokens = [tokenizer.token_to_id(token) for token in start_prompt.split()]
+text_gen_callback = TopKTextGenerator(
+    NUM_TOKENS_GENERATED, start_tokens, tokenizer
+)
+
+model = create_model()
+
+model.fit(
```

Comment:
Add a section for evaluation / inference and a conclusion section summarizing what was learned from the example.

Reply:
added inference section and concluding paragraph